pennylane_lightning/lightning_qubit/_state_vector.py

```diff
@@ -0,0 +1,361 @@
+            device_wires (Wires): wires that get initialized in the state
+        """
+
+        if isinstance(state, self._qubit_state.__class__):
+            state_data = allocate_aligned_array(state.size, np.dtype(self.dtype), True)
+            self._qubit_state.getState(state_data)
+            state = state_data
+
+        ravelled_indices, state = self._preprocess_state_vector(state, device_wires)
+
+        # translate to wire labels used by device
+        output_shape = [2] * self.num_wires
+
+        if len(device_wires) == self.num_wires and Wires(sorted(device_wires)) == device_wires:
+            # Initialize the entire device state with the input state
+            state = np.reshape(state, output_shape).ravel(order="C")
+            self._qubit_state.UpdateData(state)
+            return
+
+        self._qubit_state.setStateVector(ravelled_indices, state)
+
+    def _apply_basis_state(self, state, wires):
+        """Initialize the state vector in a specified computational basis state.
+
+        Args:
+            state (array[int]): computational basis state of shape ``(wires,)``
+                consisting of 0s and 1s.
+            wires (Wires): wires that the provided computational state should be
+                initialized on
+
+        Note: This function does not support broadcasted inputs yet.
+        """
+        num = self._get_basis_state_index(state, wires)
+        self._create_basis_state(num)
+
+    def _apply_lightning_controlled(self, operation):
+        """Apply an arbitrary controlled operation to the state tensor.
+
+        Args:
+            operation (~pennylane.operation.Operation): operation to apply
+
+        Returns:
+            array[complex]: the output state tensor
+        """
+        state = self.state_vector
+
+        basename = operation.base.name
+        if basename == "Identity":
+            return
+        method = getattr(state, f"{basename}", None)
+        control_wires = list(operation.control_wires)
+        control_values = operation.control_values
+        target_wires = list(operation.target_wires)
+        if method is not None:  # apply n-controlled specialized gate
+            inv = False
+            param = operation.parameters
+            method(control_wires, control_values, target_wires, inv, param)
+        else:  # apply gate as an n-controlled matrix
+            method = getattr(state, "applyControlledMatrix")
+            target_wires = self.wires.indices(operation.target_wires)
+            try:
+                method(
+                    qml.matrix(operation.base),
+                    control_wires,
+                    control_values,
+                    target_wires,
+                    False,
+                )
+            except AttributeError:  # pragma: no cover
+                # To support older versions of PL
+                method(operation.base.matrix, control_wires, control_values, target_wires, False)
```

Comment:
I don't think we need to try and support versions of pennylane that old. Especially because our pl version support is now blocked by other things.